docs/environment.md
# Environments

## Environment Definition File

An environment can be defined in a json file like this (`build/joomla/default.json`):

```json
{
  "name": "default",
  "server": {
    "type":"nginx",
    "offset": "UTC"
  },
  "database": {
    "driver": "mysql",
    "version": "latest",
    "host": "localhost",
    "port": "3306",
    "name": "joomla_test",
    "user": "sqladmin",
    "password": "sqladmin",
    "rootPassword": "root",
    "prefix": "jos_"
  },
  "joomla": {
    "version": "latest",
    "sampleData": "",
    "cache": {
      "enabled": false,
      "time": 15,
      "handler": "file"
    },
    "debug": {
      "system": true,
      "language": true
    },
    "meta": {
      "description": "Test installation",
      "keywords": "",
      "showVersion": true,
      "showTitle": true,
      "showAuthor": true
    },
    "sef": {
      "enabled": false,
      "rewrite": false,
      "suffix": false,
      "unicode": false
    },
    "feeds": {
      "limit": 10,
      "email": "author"
    },
    "session": {
      "lifetime": "15",
      "handler": "database"
    }
  }
}
```

## Environment Variables

All values from the environment definition file can be overridden by environment variables.
These variables can be set directly or in an `.env` file. 
The `.env` file in the project's root directory is loaded automatically if present.
Example (`build/joomla/.env`):

```dotenv
JCLI_NAME=default
JCLI_SERVER_TYPE=nginx
JCLI_SERVER_OFFSET=UTC
JCLI_DATABASE_DRIVER=mysql
JCLI_DATABASE_VERSION=latest
JCLI_DATABASE_HOST=localhost
JCLI_DATABASE_PORT=3306
JCLI_DATABASE_NAME=joomla_test
JCLI_DATABASE_USER=sqladmin
JCLI_DATABASE_PASSWORD=sqladmin
JCLI_DATABASE_ROOTPASSWORD=root
JCLI_DATABASE_PREFIX=jos_
JCLI_JOOMLA_VERSION=latest
JCLI_JOOMLA_SAMPLEDATA=
JCLI_JOOMLA_CACHE_ENABLED=0
JCLI_JOOMLA_CACHE_TIME=15
JCLI_JOOMLA_CACHE_HANDLER=file
JCLI_JOOMLA_DEBUG_SYSTEM=1
JCLI_JOOMLA_DEBUG_LANGUAGE=1
JCLI_JOOMLA_META_DESCRIPTION=Test installation
JCLI_JOOMLA_META_KEYWORDS=
JCLI_JOOMLA_META_SHOWVERSION=1
JCLI_JOOMLA_META_SHOWTITLE=1
JCLI_JOOMLA_META_SHOWAUTHOR=1
JCLI_JOOMLA_SEF_ENABLED=0
JCLI_JOOMLA_SEF_REWRITE=0
JCLI_JOOMLA_SEF_SUFFIX=0
JCLI_JOOMLA_SEF_UNICODE=0
JCLI_JOOMLA_FEEDS_LIMIT=10
JCLI_JOOMLA_FEEDS_EMAIL=author
JCLI_JOOMLA_SESSION_LIFETIME=15
JCLI_JOOMLA_SESSION_HANDLER=database
```


CHANGELOG.md
# [PROJECT_NAME] Changelog

## Version HEAD -> master

    2019-09-06 [fd27a59] Missed closing div tag (Niels Braczek)
    2019-09-06 [e43ffb3] Fix filename (Niels Braczek)
    2019-09-06 [a029c49] Minor refactoring (Niels Braczek)
    2019-09-06 [7e3d16c] Fix UML injection for classes without inheritance (Niels Braczek)
    2019-09-06 [3d40e65] Diagram filenames are lowercase (Niels Braczek)
    2019-09-06 [600e9af] Call the new command (Niels Braczek)
    2019-09-06 [5b9d493] Speed up SVG generation (Niels Braczek)
    2019-09-06 [581b93e] No longer need for temporary directory (Niels Braczek)
    2019-09-06 [0fa58af] Activate progress option (Niels Braczek)
    2019-09-06 [1771aec] After moving cleanup, only scanned files are relevant (Niels Braczek)
    2019-09-06 [d24a167] Cleanup was in the wrong place :( (Niels Braczek)
    2019-09-06 [6c907d8] Always include skin.uml (Niels Braczek)
    2019-09-06 [dcad3ca] Report each file only once (Niels Braczek)
    2019-09-06 [d63d9b3] Proper use of temporary directory (Niels Braczek)
    2019-09-06 [f21c61a] Only keep relevant files (Niels Braczek)
    2019-09-06 [d27906e] Refactor command (Niels Braczek)
    2019-09-06 [bdb28d7] Ignore anonymous classes (2) (Niels Braczek)
    2019-09-06 [33ce94e] Ignore anonymous classes (Niels Braczek)
    2019-09-06 [2bcaf9e] Control SVG generation, update documentation (Niels Braczek)
    2019-09-06 [5138ea3] Refactored UML Documentation (Niels Braczek)
    2019-09-04 [0e0fd2c] Predefine class diagrams for J1.5, J2.5 and J3 (Niels Braczek)
    2019-09-04 [37411ae] Predefine class diagrams for PHP and J3.8 (Niels Braczek)
    2019-09-03 [db77e98] Use classmap for UML rendering (Niels Braczek)
    2019-09-03 [b228800] Fix class mapping (Niels Braczek)
    2019-09-03 [f19137d] Fix local setup (Niels Braczek)
    2019-09-03 [b9caf2e] Provide Joomla! target version (Niels Braczek)
    2019-09-03 [6c57c68] Copy a local Joomla for testing (Niels Braczek)
    2019-09-02 [46c0f96] Choose predefined class diagrams depending on version (Niels Braczek)
    2019-09-02 [12feedc] Add predefined class diagrams for Joomla! 3.8 (Niels Braczek)
    2019-09-02 [0d1161f] -k option needs no value (2) (Niels Braczek)
    2019-09-02 [578d350] -k option needs no value (Niels Braczek)
    2019-09-02 [c77b1d4] Add missing newline (Niels Braczek)
    2019-09-02 [6b2bade] Recursively also copy includes (Niels Braczek)
    2019-09-02 [abc93bd] Generate diagram for unknown classes (Niels Braczek)
    2019-09-02 [41317f3] Make keepSources optional (Niels Braczek)
    2019-09-02 [02552e2] Copy predefined classes only when needed (2) (Niels Braczek)
    2019-09-02 [8fac255] Copy predefined classes only when needed (Niels Braczek)
    2019-09-02 [9d5a38e] Set default to discard UML sources (Niels Braczek)
    2019-09-02 [439116f] Remove unneeded diagrams (Niels Braczek)
    2019-09-02 [fe6e579] Provide Joomla! class diagrams (Niels Braczek)
    2019-09-02 [56581f6] Optionally keep UML sources (Niels Braczek)
    2019-09-02 [33f3103] Fix container paths (Niels Braczek)
    2019-09-02 [d71afed] Fix container paths (Niels Braczek)
    2019-09-02 [c63036b] Fix container paths (Niels Braczek)
    2019-09-02 [e7810b4] Run docker exec with local user (Niels Braczek)
    2019-09-02 [d346107] Stop containers after tests (Niels Braczek)
    2019-09-02 [6342c65] Prepare bootstrap for unit tests (Niels Braczek)
    2019-09-02 [6268c9d] Prepare bootstrap for unit tests (Niels Braczek)
    2019-09-02 [ac75aa9] More work on docs (Niels Braczek)
    2019-09-01 [8a2f998] Add some documentation stubs (Niels Braczek)
    2019-09-01 [02a18f7] Remove no longer needed directories (Niels Braczek)
    2019-08-31 [7a92e20] Remove debug output (Niels Braczek)
    2019-08-31 [2d0ce9b] Add PHPDocumentor2 (Niels Braczek)
    2019-08-30 [94fc507] Add PHPDocumentor2 (WIP) (Niels Braczek)
    2019-08-30 [a996dba] Fix small isuues (Niels Braczek)
    2019-08-30 [c7f5dbe] Fix small isuues (Niels Braczek)
    2019-08-30 [9322fb5] Extract apigen method (Niels Braczek)
    2019-08-30 [bc7c959] Tweak skins (Niels Braczek)
    2019-08-30 [24a2273] Remove JS implementation of PlantUML (Niels Braczek)
    2019-08-30 [56d7602] Use pre-rendered sequence diagrams (Niels Braczek)
    2019-08-30 [93619cb] Run Apigen in a docker container due to incompatibilities with environment (Niels Braczek)
    2019-08-29 [b49da8a] Remove Apigen - incompatible with current PHP (Niels Braczek)
    2019-08-29 [123abe8] Add Apigen (Niels Braczek)
    2019-08-29 [91ecc12] Fix UML generation (Niels Braczek)
    2019-08-29 [7b87375] Add Codebrowser (Niels Braczek)
    2019-08-29 [76b68e6] Add Joomla code style rules (Niels Braczek)
    2019-08-29 [27c4301] Look for standard in the right place (Niels Braczek)
    2019-08-29 [4ff2561] Add phpcpd (Niels Braczek)
    2019-08-29 [11ea51f] Adjust phpmd parameters (Niels Braczek)
    2019-08-29 [0eb1b34] ... redo (Niels Braczek)
    2019-08-29 [4dcf71b] ... also here (Niels Braczek)
    2019-08-29 [e5dc479] buildTemplate now points to joomla-cli's own build directory (Niels Braczek)
    2019-08-29 [9f65094] Use nginx proxy by jwilder (Niels Braczek)
    2019-08-29 [3bf0108] Added pdepend and phpmd (Niels Braczek)
    2019-08-29 [95e653d] Let echo know about 'verbose' (Niels Braczek)
    2019-08-29 [76298db] echo now is verbosity sensitive (Niels Braczek)
    2019-08-29 [2cc47d0] Removed duplicate directory (Niels Braczek)
    2019-08-29 [729739c] Separate docker and instance templates (Niels Braczek)
    2019-08-29 [0582112] Remove basedir from output (Niels Braczek)
    2019-08-29 [d23a83d] Added missing return to closure (Niels Braczek)
    2019-08-29 [af3af74] Fixed destination path (Niels Braczek)
    2019-08-29 [16c00b1] Added ablity to retrieve directory (Niels Braczek)
    2019-08-29 [0ffbd0e] Allow dot-notation for array access (Niels Braczek)
    2019-08-29 [38343c2] Only delete existing files (Niels Braczek)

src/Settings.php
<?php
/**
 * GreenCape Joomla Command Line Interface
 *
 * MIT License
 *
 * Copyright (c) 2012-2019, Niels Braczek <nbraczek@bsds.de>. All rights reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * @package         GreenCape\JoomlaCLI
 * @subpackage      Command
 * @author          Niels Braczek <nbraczek@bsds.de>
 * @copyright   (C) 2012-2019 GreenCape, Niels Braczek <nbraczek@bsds.de>
 * @license         http://opensource.org/licenses/MIT The MIT license (MIT)
 * @link            http://greencape.github.io
 * @since           File available since Release __DEPLOY_VERSION__
 */

namespace GreenCape\JoomlaCLI;

use GreenCape\JoomlaCLI\Utility\Array_;
use GreenCape\JoomlaCLI\Utility\Xml;

/**
 * Retrieve settings
 *
 * @since       Class available since Release __DEPLOY_VERSION__
 */
class Settings
{
    /**
     * @var string
     */
    private $projectName;

    public function __construct(string $projectName)
    {
        $this->projectName = $projectName;
    }

    /**
     * Get the environment settings
     *
     * @param  string  $actualXmlFile  The name of the XML file containing the environment definition
     * @param  string  $defaultDir     The directory containing the default environment definition (`default.xml`)
     *
     * @return array
     */
    public function environment(string $actualXmlFile, string $defaultDir): array
    {
        $target = basename($actualXmlFile, '.xml');

        $environment = [
            'name'     => $target,
            'server'   => [
                'type'   => 'nginx',
                'offset' => 'UTC',
                'tld'    => 'dev',
            ],
            'php'      => [
                'version' => 'latest',
            ],
            'cache'    => [
                'enabled' => 0,
                'time'    => 15,
                'handler' => 'file',
            ],
            'debug'    => [
                'system'   => 1,
                'language' => 1,
            ],
            'meta'     => [
                'description' => "Test installation for {$this->projectName} on Joomla! \${version}",
                'keywords'    => "{$this->projectName} Joomla Test",
                'showVersion' => 1,
                'showTitle'   => 1,
                'showAuthor'  => 1,
            ],
            'sef'      => [
                'enabled' => 0,
                'rewrite' => 0,
                'suffix'  => 0,
                'unicode' => 0,
            ],
            'session'  => [
                'lifetime' => 15,
                'handler'  => 'database',
            ],
            'joomla'   => [
                'version'    => 'latest',
                'sampleData' => 'data',
            ],
            'database' => [
                'driver' => 'mysqli',
                'name'   => 'joomla_test',
                'prefix' => '${target}_',
            ],
            'feeds'    => [
                'limit' => 10,
                'email' => 'author',
            ],
        ];

        $environment = Array_::merge(
            $environment,
            Xml::xmlProperty($defaultDir . '/default.xml', false, true)
        );
        $environment = Array_::merge(
            $environment,
            Xml::xmlProperty($actualXmlFile, false, true)
        );

        $environment['meta']['description'] = str_replace(
            '${version}',
            $environment['joomla']['version'],
            $environment['meta']['description']
        );
        $environment['database']['prefix']  = str_replace(
            '${target}',
            $environment['name'],
            $environment['database']['prefix']
        );

        if (in_array($environment['database']['driver'], ['mysqli', 'pdomysql'])) {
            $environment['database']['engine'] = 'mysql';
        } else {
            $environment['database']['engine'] = $environment['database']['driver'];
        }

        $database = $this->defaultDatabase($defaultDir);

        // Get the database info - use global values, if not provided with local environment
        $environment['database']['version']      = $environment['database']['version'] ?? $database[$environment['database']['engine']]['version'];
        $environment['database']['name']         = $environment['database']['name'] ?? $database[$environment['database']['engine']]['name'];
        $environment['database']['user']         = $environment['database']['user'] ?? $database[$environment['database']['engine']]['user'];
        $environment['database']['password']     = $environment['database']['password'] ?? $database[$environment['database']['engine']]['password'];
        $environment['database']['rootPassword'] = $environment['database']['rootPassword'] ?? $database[$environment['database']['engine']]['rootPassword'];

        return $environment;
    }

    /**
     * Load database environment, if provided, using default values for keys not defined in `$xmlFile`
     *
     * @param  string  $defaultDir  The directory containing the default database setting (`database.xml`)
     *
     * @return array
     */
    public function defaultDatabase(string $defaultDir): array
    {
        $database = [
            'mysql'      => [
                'version'      => 'latest',
                'host'         => 'localhost',
                'port'         => '3306',
                'name'         => 'joomla_test',
                'user'         => 'db_user',
                'password'     => 'db_pass',
                'rootPassword' => '',
            ],
            'postgresql' => [
                'version'      => 'latest',
                'host'         => 'localhost',
                'port'         => '5432',
                'name'         => 'joomla_test',
                'user'         => 'db_user',
                'password'     => 'db_pass',
                'rootPassword' => '',
            ],
        ];

        if (file_exists($defaultDir . '/database.xml')) {
            $database = Array_::merge(
                $database,
                Xml::xmlProperty($defaultDir . '/database.xml', false, true)
            );
        }

        $database['mysql']['passwordOption'] = empty($database['mysql']['rootPassword'])
            ? ''
            : "-p'{$database['mysql']['rootPassword']}'";

        return $database;
    }
}
